quinn/src/runtime.rs

```diff
@@ -87,78 +76,161 @@ pub trait AsyncUdpSocket: Send + Sync + Debug + 'static {
     }
 }
 
-/// An object polled to detect when an associated [`AsyncUdpSocket`] is writable
+/// An object for asynchronously writing to an associated [`AsyncUdpSocket`].
 ///
-/// Any number of `UdpPoller`s may exist for a single [`AsyncUdpSocket`]. Each `UdpPoller` is
-/// responsible for notifying at most one task when that socket becomes writable.
-pub trait UdpPoller: Send + Sync + Debug + 'static {
-    /// Check whether the associated socket is likely to be writable
-    ///
-    /// Must be called after [`AsyncUdpSocket::try_send`] returns [`io::ErrorKind::WouldBlock`] to
-    /// register the task associated with `cx` to be woken when a send should be attempted
-    /// again. Unlike in [`Future::poll`], a [`UdpPoller`] may be reused indefinitely no matter how
-    /// many times `poll_writable` returns [`Poll::Ready`].
-    fn poll_writable(self: Pin<&mut Self>, cx: &mut Context) -> Poll<io::Result<()>>;
+/// Any number of [`UdpSender`]s may exist for a single [`AsyncUdpSocket`]. Each [`UdpSender`] is
+/// responsible for notifying at most one task for send readiness.
+pub trait UdpSender: Send + Sync + Debug + 'static {
+    /// Send a UDP datagram, or register to be woken if sending may succeed in the future.
+    ///
+    /// Usually implementations of this will poll the socket for writability before trying to
+    /// write to them, and retry both if writing fails.
+    ///
+    /// Quinn will create multiple [`UdpSender`]s, one for each task it's using it from. Thus it's
+    /// important to poll the underlying socket in a way that doesn't overwrite wakers.
+    ///
+    /// A single [`UdpSender`] will be re-used, even if `poll_send` returns `Poll::Ready` once,
+    /// unlike [`Future::poll`], so calling it again after readiness should not panic.
+    fn poll_send(
+        self: Pin<&mut Self>,
+        transmit: &Transmit,
+        cx: &mut Context,
+    ) -> Poll<io::Result<()>>;
+
+    /// Maximum number of datagrams that a [`Transmit`] may encode.
+    fn max_transmit_segments(&self) -> usize {
+        1
+    }
+
+    /// Try to send a UDP datagram, if the socket happens to be write-ready.
+    ///
+    /// This may fail with [`io::ErrorKind::WouldBlock`], if the socket is currently full.
+    ///
+    /// The quinn endpoint uses this function when sending
+    ///
+    /// - A version negotiation response due to an unknown version
+    /// - A `CLOSE` due to a malformed or unwanted connection attempt
+    /// - A stateless reset due to an unrecognized connection
+    /// - A `Retry` packet due to a connection attempt when `use_retry` is set
+    ///
+    /// If sending in these cases fails, a well-behaved peer will re-try. Thus it's fine
+    /// if we drop datagrams sometimes with this function.
+    fn try_send(self: Pin<&mut Self>, transmit: &Transmit) -> io::Result<()>;
```

Comment:
Would it be simpler to omit this and call `poll_send` with a no-op waker?

Reply:
Unfortunately, you don't have a dependency that provides a no-op waker yet. There's a `std::task::Waker::noop()`, but it's only stable since rust 1.85, which is far younger than your MSRV 1.71, so I've inlined the implementation.